gov/proposal.go

```diff
@@ -0,0 +1,164 @@
+package gov
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/maticnetwork/heimdall/gov/types"
+	hmTypes "github.com/maticnetwork/heimdall/types"
+)
+
+// SubmitProposal create new proposal given a content
+func (keeper Keeper) SubmitProposal(ctx sdk.Context, content types.Content) (types.Proposal, sdk.Error) {
+	if !keeper.router.HasRoute(content.ProposalRoute()) {
+		return types.Proposal{}, types.ErrNoProposalHandlerExists(keeper.codespace, content)
+	}
+
+	// Execute the proposal content in a cache-wrapped context to validate the
+	// actual parameter changes before the proposal proceeds through the
+	// governance process. State is not persisted.
+	cacheCtx, _ := ctx.CacheContext()
+	handler := keeper.router.GetRoute(content.ProposalRoute())
+	if err := handler(cacheCtx, content); err != nil {
```

Comment:
why are we calling handler here on submit proposal? 

In my understand, handler should be called only in end-blocker once required votes are received. handler will update the params.

Reply:
It validates if the handler will actually execute when the proposal succeeds. The state change is not be persisted since it is cache context.

FYI: This file is borrowed from cosmos-sdk `x/gov` package.

Reply:
Got it. Thanks.